packages/wallet/ui/src/bridge.jsx

```diff
@@ -0,0 +1,144 @@
+  loaded: 'walletBridgeLoaded',
+  opened: 'walletBridgeOpened',
+});
+
+/**
+ * Install a dApp "connection" where messages posted from
+ * the dApp are forwarded to localStorage and vice versa.
+ *
+ * @param {{
+ *   addEventListener: typeof window.addEventListener,
+ *   parentPost: typeof window.postMessage,
+ *   t0: number,
+ * }} io
+ */
+const installDappConnection = ({ addEventListener, parentPost, t0 }) => {
+  /** @type { string } */
+  let origin;
+
+  /** @type {(key: string, value: string) => void} */
+  let setItem;
+  /** @type {string[]} */
+  const buffer = [];
+
+  console.debug('bridge: installDappConnection');
+
+  parentPost({ type: BridgeProtocol.loaded }, '*');
+  parentPost({ type: BridgeProtocol.opened }, '*');
+
+  let outIx = 0; // TODO: persist index in storage
+  /** @param {Record<string, any>} payload */
+
+  const { stringify, parse } = JSON;
+
+  addEventListener('message', ev => {
+    // console.debug('bridge: handling message:', ev.data);
+    if (
+      !ev.data ||
+      typeof ev.data.type !== 'string' ||
+      !ev.data.type.startsWith('CTP_')
+    ) {
+      return;
+    }
+
+    if (origin === undefined) {
+      // First-come, first-serve.
+      // console.debug('bridge: setting origin to', origin);
+      origin = ev.origin;
+    }
+    if (setItem) {
+      console.debug('bridge: message from dapp->localStorage', origin, ev.data);
+      setItem(stringify(['out', origin, t0, outIx]), stringify(ev.data));
+    } else {
+      console.debug('bridge: message from dapp->buffer', origin, ev.data);
+      buffer.push(harden(ev.data));
+    }
+  });
+
+  return harden({
+    /** @param {typeof window.localStorage} storage */
+    connectStorage: storage => {
+      addEventListener('storage', ev => {
+        // console.debug('from storage', origin, ev.key, ev.newValue);
+        if (!ev.key || !ev.newValue) {
+          return;
+        }
+        const [tag, targetOrigin, targetT, _keyIx] = JSON.parse(ev.key); // or throw
+        if (tag !== 'in' || targetOrigin !== origin || targetT !== t0) {
+          return;
+        }
+        const payload = parse(ev.newValue); // or throw
+        storage.removeItem(ev.key);
+        console.debug('bridge: storage -> message to dapp', origin, payload);
+        parentPost(payload, '*');
+      });
+      setItem = (key, value) => {
+        storage.setItem(key, value);
+        outIx += 1;
+      };
+      if (buffer.length) {
+        console.debug('sending', buffer.length, 'queued messages from', origin);
+        while (buffer.length) {
+          setItem(
+            stringify(['out', origin, t0, outIx]),
+            stringify(buffer.shift()),
+          );
+        }
+      }
+    },
+  });
+};
+
+const sameParent = () => {
+  const me = window;
+  const { parent } = window;
+  if (me === parent) {
+    // eslint-disable-next-line no-debugger
+    debugger;
+    throw Error('window.parent === parent!!!');
+  }
+  return false;
```

Comment:
as to why: @michaelfig wrote Jun 28:
> it’s just to prevent infinite regress when the bridge is loaded in its own web page.

Given that, the test is probably applied at the wrong place. It probably belongs when we first wire up the top level event handlers, not on each message.